### DISCUSSION TITLE
Having problems with query using where on fields of type DateTime

### DISCUSSION THREAD

**Discussion by rhscjohn-dev** [prisma/prisma#8956] [Q&A] [answered]
Newbie to Prisma 2.  SQLite3 database.  

Prisma Model of table in question.  Model wad generated via introspection of the db. 
```prisma
model EPG_EVENT {
  oid                  Int       @id @default(autoincrement())
  title                String
  subtitle             String
  description          String
  start_time           DateTime
  end_time             DateTime
  channel_oid          Int
  unique_id            String
  rating               String
  original_air_date    DateTime?
  season               Int
  episode              Int
  dvb_service_event_id Int?
  dvb_table_version    Int?
  genres               String?
  first_run            String?
  quality              String?
  aspect               String?
  audio                String?
  star_rating          String?
  cast_member          String?
  crew                 String?
  md5                  String?
  significance         String?

  @@index([start_time], name: "IX_start_time")
  @@index([channel_oid], name: "IX_channel_oid")
}
```

Here is the Prisma `query.
```ts 
let startDate = new Date()
  startDate.setDate(startDate.getDate() + 1)  // Add 1 day to today's date
  startDate.setHours(0)
  startDate.setMinutes(0)
  startDate.setSeconds(0)
  startDate.setMilliseconds(0)
  let endDate = new Date(startDate) // enddate is 24 hours from startdate
  endDate.setHours(23)
  endDate.setMinutes(59)
  endDate.setSeconds(59)
  console.log(`startDate: ${startDate.toISOString()} endDate: ${endDate.toISOString()}`)
  const channelEPGEvents = await prisma.ePG_EVENT.findMany({
    where: {
      channel_oid: { in: [7156] },
      start_time: {
        // gte: new Date(startDate),
        // lte: new Date(endDate),
        gte: startDate.toISOString(),
        lte: endDate.toISOString()
      }
    },
    select: {
      oid: true,
      channel_oid: true,
      title: true,
      subtitle: true,
      description: true,
      start_time: true,
      season: true,
      episode: true,
      first_run: true,
      significance: true
    }
  })
  console.log(channelEPGEvents)
```

I have tried two different ways of setting the values for the field: start_time with the same result.  I have also tried Prisma Studio and also got the same result of an empty array.

Console log output.
```
*** all epgEvents for a list of channels with Channel Group Name of test
startDate: 2021-08-27T04:00:00.000Z endDate: 2021-08-28T03:59:59.000Z
Query: SELECT `main`.`EPG_EVENT`.`oid`, `main`.`EPG_EVENT`.`channel_oid`, `main`.`EPG_EVENT`.`title`, `main`.`EPG_EVENT`.`subtitle`, `main`.`EPG_EVENT`.`description`, `main`.`EPG_EVENT`.`start_time`, `main`.`EPG_EVENT`.`season`, `main`.`EPG_EVENT`.`episode`, `main`.`EPG_EVENT`.`first_run`, `main`.`EPG_EVENT`.`significance` FROM `main`.`EPG_EVENT` WHERE (`main`.`EPG_EVENT`.`channel_oid` IN (?) AND `main`.`EPG_EVENT`.`start_time` >= 
? AND `main`.`EPG_EVENT`.`start_time` <= ?) LIMIT ? OFFSET ?
Duration: 1ms
[]
```

The documentation is pretty poor when it comes to fields of type DateTime and how to use them filter(s).

Any suggestions would be appreciated.

**Comment by ryands17** [comment]
@rhscjohn-dev 👋 

This is the correct way of querying with dates. Also the dates are stored in UTC in your database so make sure you're taking that into account as well.

**Comment by rhscjohn-dev** [comment]
Still can not get Prisma to query on a DateTime field.

Here are all of the combinations I have tried.  They either say throw an error: 
```
Argument start_time: Got invalid value
{
  gte: '2021-08-28 04:00:00',
  lte: '2021-08-29 03:59:59'
}
```
or return an empty array.

Prisma Client code.
``` 
const channelEPGEvents = await prisma.ePG_EVENT.findMany({
    where: {
      channel_oid: { in: [7156] },
      start_time: {
        // gte: new Date(startDate),
        // lte: new Date(endDate),
        // gte: startDate.toISOString(),
        // lte: endDate.toISOString()
        // gte: startDate.toISOString().replace(/T/, ' ').split('.')[0],  // invalid value
        // lte: endDate.toISOString().replace(/T/, ' ').split('.')[0]     // invalid value
        // gte: "2021-08-28T04:00:00",  // invalid value
        // lte: "2021-08-29T03:59:59"  // invalid value
        // gte: "2021-08-28T04:00:00Z", // []
        // lte: "2021-08-29T03:59:59Z"  // []
        // gte: "2021-08-28 04:00:00Z",  // invalid value
        // lte: "2021-08-29 03:59:59Z"   // invalid value
        // gte: new Date("2021-08-28 04:00:00Z"),  // []
        // lte: new Date("2021-08-29 03:59:59Z")   // []
        // gte: new Date("2021-08-28 04:00:00.000Z"),  // []
        // lte: new Date("2021-08-29 03:59:59.000Z")   // []
        // gte: new Date("2021-08-28 04:00:00.000"),  // []
        // lte: new Date("2021-08-29 03:59:59.000")   // []
        gte: "2021-08-28 04:00:00",  // invalid value
        lte: "2021-08-29 03:59:59"   // invalid value
      }
    },
    select: {
      oid: true,
      channel_oid: true,
      title: true,
      subtitle: true,
      description: true,
      start_time: true,
      season: true,
      episode: true,
      first_run: true,
      significance: true
    }
  })
```

I do not have this problem when using Knex.  So I went to an independent third party program - DB Browser for SQLite and composed the SQLite statement:
```
SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= "2021-08-28 04:00:00" AND start_time <= "2021-08-29 03:59:59"
```

The statement executed without errors and returned 30 rows of data.  A screen shot  of the SQL statement, result table and report is below.

![SQL Screenshot 2021-08-27 135628](https://user-images.githubusercontent.com/60523782/131170663-6a798b0d-3d2c-4acf-a2c0-60e18c5d3867.png)

**Reply by janpio** [reply]
Now that you have a working query, can you compare it to the one Prisma generated and find the differences (besides that Prisma explicitly requests the fields one by one)? Is Prisma generating an invalid query in any way? If not, this might very well be a bug somehow that we should look into.

**Reply by ryands17** [reply]
The date string you are passing is invalid. You either need to pass a `Date` object or an ISO string like this: `2021-08-27T18:21:06.276Z`.

**Reply by rhscjohn-dev** [reply]
Prisma throws an error, so no query is generated to look at.

The error response is
```
ERROR Code: undefined Cause: undefined 
Invalid `prisma.ePG_EVENT.findMany()` invocation in
D:\Users\johni\Projects\hello-prisma\src\script.js:172:51

  169 endDate.setMinutes(59)
  170 endDate.setSeconds(59)
  171 console.log(`startDate: ${startDate.toISOString()} endDate: ${endDate.toISOString()}`)     
→ 172 const channelEPGEvents = await prisma.ePG_EVENT.findMany({
        where: {
          channel_oid: {
            in: [
              7156
            ]
          },
          start_time: {
            gte: '2021-08-28 04:00:00',
            lte: '2021-08-29 03:59:59'
          }
          ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
        },
        select: {
          oid: true,
          channel_oid: true,
          title: true,
          subtitle: true,
          description: true,
          start_time: true,
          season: true,
          episode: true,
          first_run: true,
          significance: true
        }
      })

Argument start_time: Got invalid value
{
  gte: '2021-08-28 04:00:00',
  lte: '2021-08-29 03:59:59'
}
on prisma.findManyEPG_EVENT. Provided Json, expected DateTimeFilter or DateTime.
type DateTimeFilter {
  equals?: DateTime
  in?: List<DateTime>
  notIn?: List<DateTime>
  lt?: DateTime
  lte?: DateTime
  gt?: DateTime
  gte?: DateTime
  not?: DateTime | NestedDateTimeFilter
}`

**Reply by rhscjohn-dev** [reply]
> The date string you are passing is invalid. You either need to pass a `Date` object or an ISO string like this: `2021-08-27T18:21:06.276Z`.

I have tried:
`// gte: "2021-08-28T04:00:00Z", // []
        // lte: "2021-08-29T03:59:59Z"  // []`
and 
`gte: new Date("2021-08-28 04:00:00Z"),  // []
        // lte: new Date("2021-08-29 03:59:59Z")   // []`
and both return an empty array.

The Query statement generated by prisma is
```
Query: SELECT `main`.`EPG_EVENT`.`oid`, `main`.`EPG_EVENT`.`channel_oid`, `main`.`EPG_EVENT`.`title`, `main`.`EPG_EVENT`.`subtitle`, `main`.`EPG_EVENT`.`description`, `main`.`EPG_EVENT`.`start_time`, `main`.`EPG_EVENT`.`season`, `main`.`EPG_EVENT`.`episode`, `main`.`EPG_EVENT`.`first_run`, 
`main`.`EPG_EVENT`.`significance` FROM `main`.`EPG_EVENT` WHERE (`main`.`EPG_EVENT`.`channel_oid` IN (?) AND `main`.`EPG_EVENT`.`start_time` >= ? AND `main`.`EPG_EVENT`.`start_time` <= ?) LIMIT 
? OFFSET ?
Duration: 1ms
``` 

And the result is an empty array

**Reply by janpio** [reply]
How is that generated query different from the query you manually created before and that returned the wanted data?

**Reply by rhscjohn-dev** [reply]
My generated statement has the value(s) hard coded. and is using a SELECT *
`SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= "2021-08-28 04:00:00" AND start_time <= "2021-08-29 03:59:59"
`
The prisma statement has ? for the values and Select fields.
``` 
Query: SELECT `main`.`EPG_EVENT`.`oid`, `main`.`EPG_EVENT`.`channel_oid`, `main`.`EPG_EVENT`.`title`, `main`.`EPG_EVENT`.`subtitle`, `main`.`EPG_EVENT`.`description`, `main`.`EPG_EVENT`.`start_time`, `main`.`EPG_EVENT`.`season`, `main`.`EPG_EVENT`.`episode`, `main`.`EPG_EVENT`.`first_run`, 
`main`.`EPG_EVENT`.`significance` FROM `main`.`EPG_EVENT` WHERE (`main`.`EPG_EVENT`.`channel_oid` IN (?) AND `main`.`EPG_EVENT`.`start_time` >= ? AND `main`.`EPG_EVENT`.`start_time` <= ?) LIMIT 
? OFFSET ?
Duration: 1ms
```
Other than the ? and Select fields, I do not see a difference.

**Reply by ryands17** [reply]
Seems like a UTC issue then. Run your script like this:

```
TZ=utc node script.js
```

Most likely your date is being converted to UTC which is why you're not getting the correct results.

**Reply by rhscjohn-dev** [reply]
I'll give that a try but my date range for start-time in that table starts at 2021-08-26 16:30:00 and ends at 2021-09-09 23:30:30.

Tried your suggestion and was unsuccessful
`D:\Users\johni\Projects\hello-prisma>TZ=utc node script.js
'TZ' is not recognized as an internal or external command,
operable program or batch file.`

**Reply by janpio** [reply]
On Windows that would look more like this:
```
set TZ=utc
node script.js
```
(2 separate commands)

**Reply by ThatAnonyG** [reply]
That did not work. I have the same issue.

**Reply by rhscjohn-dev** [reply]
Prisma will only accept type Number in SQlite3 DateTime field.  If your DateTime field contains a string, Prisma will choke.  I abounded Prisma and went back to Knex.  Prisma had no interest in accepting a string in a DateTime field even though a string is valid in SQLite3.

**Reply by janpio** [reply, marked as answer]
If you are still interested in that, please open a feature request with a clear description of what should work, but doesn't. Thanks.

**Reply by ThatAnonyG** [reply]
> Prisma will only accept type Number in SQlite3 DateTime field. If your DateTime field contains a string, Prisma will choke. I abounded Prisma and went back to Knex. Prisma had no interest in accepting a string in a DateTime field even though a string is valid in SQLite3.

By number do you mean UNIX timestamps? And is that lack of support for JS Date object too? I tried JS Date and ISO String none of them worked.

**Reply by ThatAnonyG** [reply]
> If you are still interested in that, please open a feature request with a clear description of what should work, but doesn't. Thanks.

I would very much like to have my code working so yeah. Currently Prisma doesn't seem to be supporting queries using ISO Date string AND JS Date object. Whenever I use that in a query Prisma replaces the Dates with `?` in the query. Something like below.

Correct query that works:
```sql
SELECT id FROM user WHERE last_login > "2022-03-28T19:10:10.216Z";
```
Prisma code:
```ts
this.prismaService.user.findMany({
  where: {
    last_login: {
      gt: lastLoginDate.toISOString(),
    },
  },
  select: { id: true },
});
```
The generated query for above code would look something like this:
```sql
SELECT id FROM user WHERE last_login > ?;
```
***Note that `?` mark!!***

**Reply by rhscjohn-dev** [reply]
> If you are still interested in that, please open a feature request with a clear description of what should work, but doesn't. Thanks.

I have no interest as I have discarded Prisma and went with Knex.

**Comment by rhscjohn-dev** [comment]
The Prisma Docs(prisma-schema-reference#datetime) for SQLite says that that the underlying data type is STRING then RFC 2822 format should be used.   Valid date and Time DataType for SQLite3 can be found at [https://www.sqlite.org/datatype3.html](url) section 2.2.   SQLite foundation Time values can be found at [https://www.sqlite.org/lang_datefunc.html](url).  

May be the format YYYY-MM-DD HH:MM:SS should not throw an error.  Or, when using an ISO date,  the Prisma engine for SQLite is not converting the ISO date to a valid Time Value for SQLite.

**Comment by rhscjohn-dev** [comment]
I know I have a good SQL statement.  So, I decides to try $queryRaw.

I first tried 
` const channelRaw = await prisma.$queryRaw`SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= "2021-08-28 04:00:00" AND start_time <= "2021-08-29 03:59:59";``

and got the correct result
`*** all epgEvents using raw
startString 2021-08-28 04:00:00 endString: 2021-08-29 03:59:59
Query: SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= "2021-08-28 04:00:00" AND start_time <= "2021-08-29 03:59:59";
Duration: 0ms
Row count:  30
`

I then tried
`
  description = `all epgEvents using raw`
  console.log(`${FgMagenta}*** ${description}${Reset}`)
  epgChannels = allChannelsGroupName.channel_group.map(row => {
    return row.channel.oid
  })

  startDate = new Date()
  startDate.setDate(startDate.getDate() + 1)  // Add 1 day to today's date
  startDate.setHours(0)
  startDate.setMinutes(0)
  startDate.setSeconds(0)
  startDate.setMilliseconds(0)
  endDate = new Date(startDate) // enddate is 24 hours from startdate
  endDate.setHours(23)
  endDate.setMinutes(59)
  endDate.setSeconds(59)
  startString = startDate.toISOString().replace(/T/, ' ').split('.')[0]
  endString = endDate.toISOString().replace(/T/, ' ').split('.')[0]
  console.log(`startString ${startString} endString: ${endString}`)
  // const channelRaw = await prisma.$queryRaw`SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= "2021-08-28 04:00:00" AND start_time <= "2021-08-29 03:59:59";`
  const channelRaw = await prisma.$queryRaw`SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= ${startString} AND start_time <= ${endString};`

  console.log('Row count: ', channelRaw.length)`

And I got the correct result.
`*** all epgEvents using raw
startString 2021-08-28 04:00:00 endString: 2021-08-29 03:59:59
Query: SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= ? AND start_time <= 
?;
Duration: 0ms
Row count:  30
`
My conclusion is that there is a bug in the Prisma Client when converting type Model Schema DateTime field  to 
 valid SQLite3 DateTime value.

**Comment by rhscjohn-dev** [comment]
I refactored the code.  The $queryRaw start_time (type DateTime) is now formatted as close as I can get it and still have the correct results.

The Prisma client code.
```
description = `24 hours of epgEvents(starting at midnight local) for a list of channels, filter on start_time using tommorrow's date`
  console.log(`${FgMagenta}*** ${description}${Reset}`)
  let startDate = new Date()
  startDate.setDate(startDate.getDate() + 1)  // Add 1 day to today's date
  startDate.setHours(0)
  startDate.setMinutes(0)
  startDate.setSeconds(0)
  startDate.setMilliseconds(0)
  let endDate = new Date(startDate) // enddate is 24 hours from startdate
  endDate.setHours(23)
  endDate.setMinutes(59)
  endDate.setSeconds(59)
  console.log(`startDate: ${startDate.toISOString()} endDate: ${endDate.toISOString()}`)
  const channelEPGEvents = await prisma.ePG_EVENT.findMany({
    select: {
      oid: true,
      title: true,
      start_time: true
    },
    where: {
      channel_oid: { in: [7156] },
      start_time: {
        gte: startDate.toISOString(),
        lte: endDate.toISOString()
      }
    },
  })
  console.log('Prisma findMany() Row count: ', channelEPGEvents.length)
  // setup raw start and end date strings
  // SELECT * FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= "2021-08-28 04:00:00" AND start_time <= "2021-08-29 03:59:59"
  const startString = startDate.toISOString().replace(/T/, ' ')
  const endString = endDate.toISOString().replace(/T/, ' ')
  console.log(`startString ${startString} endString: ${endString}`)
  const rawChannelEPGEvents = await prisma.$queryRaw`SELECT oid, title, start_time FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= ${startString} AND start_time <= ${endString};`
  console.log('Prisma $queryRaw Row count: ', rawChannelEPGEvents.length)
```

Even though the the two generated Query SELECT are comparable,  the Prisma findMany() still returns 0 records while the Prisma $queryRaw returns the correct number of records.  The only difference is the date value formatted fields is the missing 'T' in $queryRaw.

Console output.
```
*** 24 hours of epgEvents(starting at midnight local) for a list of channels, filter on start_time using tommorrow's date
startDate: 2021-08-31T04:00:00.000Z endDate: 2021-09-01T03:59:59.000Z
Query: SELECT `main`.`EPG_EVENT`.`oid`, `main`.`EPG_EVENT`.`title`, `main`.`EPG_EVENT`.`start_time` FROM `main`.`EPG_EVENT` WHERE (`main`.`EPG_EVENT`.`channel_oid` IN (?) AND `main`.`EPG_EVENT`.`start_time` >= ? AND `main`.`EPG_EVENT`.`start_time` <= ?) LIMIT ? OFFSET ?
Duration: 1ms
Prisma findMany() Row count:  0
startString 2021-08-31 04:00:00.000Z endString: 2021-09-01 03:59:59.000Z
Query: SELECT oid, title, start_time FROM EPG_EVENT WHERE channel_oid IN (7156) AND start_time >= ? AND start_time <= ?;
Duration: 1ms
Prisma $queryRaw Row count:  30
```

**Comment by NWPoul** [comment]
Also have problem with filter on Date
Seems that they fixed **gte** filter
but **lte** filter works weird
```
where: {
  startDateTime: { lte: new Date( '2023-01-31' ) },
}
```
return only one record with
startDateTime: Sun Jan 01 2023 14:09:32 GMT+0300

while on the same database
`startDateTime: {  gte: new Date( '2022-01-01' ) }`

gives many records which are from 2022

**Comment by shmulyeng** [comment]
Is there a solution for this? Basically, I see no way to use Prisma with sqlite if you need to filter by dates.

**Reply by rhscjohn-dev** [reply]
Wow.  Here it is 2024 and they still will not fix this.  I dropped Prisma back in 2021.

**Comment by thomas-rager** [comment]
Sorting by date is an elementary feature! If this is not fixed, then Prisma should officially drop support for SQLite to prevent Developers from running into issues with this database. I cannot understand why this is not treated as a critical bug.

**Reply by thesilph** [reply]
+1 issue still present

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
